starboard/android/shared/media_codec_bridge_eradicator.cc

```diff
@@ -0,0 +1,108 @@
+#include "starboard/android/shared/media_codec_bridge_eradicator.h"
+
+#include <pthread.h>
+
+#include "starboard/android/shared/jni_utils.h"
+#include "starboard/common/log.h"
+#include "starboard/common/once.h"
+
+namespace starboard {
+namespace android {
+namespace shared {
+
+SB_ONCE_INITIALIZE_FUNCTION(MediaCodecBridgeEradicator,
+                            MediaCodecBridgeEradicator::GetInstance);
+
+namespace {
+
+struct EradicateParam {
+  explicit EradicateParam(MediaCodecBridgeEradicator* eradicator,
+                          jobject j_media_codec_bridge,
+                          jobject j_reused_get_output_format_result)
+      : eradicator(eradicator),
+        j_media_codec_bridge(j_media_codec_bridge),
+        j_reused_get_output_format_result(j_reused_get_output_format_result) {}
+  MediaCodecBridgeEradicator* eradicator;
+  jobject j_media_codec_bridge;
+  jobject j_reused_get_output_format_result;
+};
+
+}  // namespace
+
+void MediaCodecBridgeEradicator::CheckEradicatingAndMaybeWait() {
+  ScopedLock scoped_lock(mutex_);
+  while (!j_media_codec_bridge_set_.empty()) {
+    condition_variable_.Wait();
+  }
+}
+
+bool MediaCodecBridgeEradicator::Eradicate(
+    jobject j_media_codec_bridge,
+    jobject j_reused_get_output_format_result) {
+  EradicateParam* param = new EradicateParam(this, j_media_codec_bridge,
+                                             j_reused_get_output_format_result);
+
+  // spawn a thread to run the release method on the MediaCodecBridge java
+  // object.
+  pthread_t thread;
+  int result = pthread_create(
+      &thread, nullptr, &MediaCodecBridgeEradicator::EradicateMediaCodecBridge,
+      param);
+  return result == 0;
+}
+
+void* MediaCodecBridgeEradicator::EradicateMediaCodecBridge(void* context) {
+  EradicateParam* param = static_cast<EradicateParam*>(context);
+  SB_DCHECK(param != nullptr);
+
+  MediaCodecBridgeEradicator* eradicator = param->eradicator;
+  jobject j_media_codec_bridge = param->j_media_codec_bridge;
+  jobject j_reused_get_output_format_result =
+      param->j_reused_get_output_format_result;
+
+  {
+    // add the j_media_codec_bridge to the set before the work is started
+    ScopedLock scoped_lock(eradicator->mutex_);
+    eradicator->j_media_codec_bridge_set_.insert(j_media_codec_bridge);
+  }
+
+  JniEnvExt* env = JniEnvExt::Get();
+
+  env->CallVoidMethodOrAbort(j_media_codec_bridge, "stop", "()V");
+  env->CallVoidMethodOrAbort(j_media_codec_bridge, "release", "()V");
+  env->DeleteGlobalRef(j_media_codec_bridge);
+
+  SB_DCHECK(j_reused_get_output_format_result);
+  env->DeleteGlobalRef(j_reused_get_output_format_result);
+
+  {
+    // the work is finished, delete the j_media_codec_bridge from the set
+    ScopedLock scoped_lock(eradicator->mutex_);
+    eradicator->j_media_codec_bridge_set_.erase(j_media_codec_bridge);
+
+    if (eradicator->j_media_codec_bridge_set_.empty()) {
+      eradicator->condition_variable_.Signal();
+    }
+  }
+
+  delete param;
```

Comment:
This can be deleted once the members are copied out above.

Reply:
sure, moved up